Geometry/Fields/ScalarFields/PieceConstantFields/cartesianField_class.f90

```diff
@@ -0,0 +1,398 @@
+    real(defReal), dimension(:,:), allocatable    :: tempMap
+    character(nameLen), dimension(:), allocatable :: mats
+    character(100), parameter :: Here = 'init (cartesianField_class.f90)'
+    
+    ! Load pitch
+    call dict % get(temp, 'pitch')
+    N = size(temp)
+
+    if (N /= 3) then
+      call fatalError(Here, 'Pitch must have size 3. Has: '//numToChar(N))
+    end if
+    self % pitch = temp
+    
+    ! Load origin
+    call dict % get(temp, 'origin')
+    N = size(temp)
+
+    if (N /= 3) then
+      call fatalError(Here, 'Origin must have size 3. Has: '//numToChar(N))
+    end if
+    origin = temp
+
+    ! Load Size
+    call dict % get(tempI, 'shape')
+    N = size(tempI)
+
+    if (N /= 3) then
+      call fatalError(Here, 'Shape must have size 3. Has: '//numToChar(N))
+    else if (any(tempI < 0)) then
+      call fatalError(Here, 'Shape contains -ve entries')
+    end if
+    self % sizeN = tempI
+
+    ! Detect reduced Z dimension
+    if (self % sizeN(3) == 0) then
+      self % sizeN(3) = 1
+      self % pitch(3) = TWO * INF
+    end if
+
+    ! Check X & Y for 0 size
+    if (any( self % sizeN == 0)) call fatalError(Here, 'Shape in X and Y axis cannot be 0.')
+
+    ! Check for invalid pitch
+    if (any(self % pitch < 10 * SURF_TOL)) then
+     call fatalError(Here, 'Pitch size must be larger than: '//numToChar( 10 * SURF_TOL))
+   end if
+
+    ! Calculate halfwidth and corner
+    self % a_bar = self % pitch * HALF - SURF_TOL
+    self % corner = origin -(self % sizeN * HALF * self % pitch)
+
+    ! Build outline box
+    call tempDict % init(4)
+    call tempDict % store('type', 'box')
+    call tempDict % store('id', 1)
+    call tempDict % store('origin', origin)
+    call tempDict % store('halfwidth', abs(self % corner - origin))
+    call self % outline % init(tempDict)
+
+    ! Construct fill array
+    ! Detect how many materials are present
+    self % nMat = dict % getSize('materials')
+
+    call dict % get(mats, 'materials')
+    
+    if (any(mats == 'all') .and. self % nMat > 1) call fatalError(Here, 'Material "all" '//&
+            'can only be used by itself and is a reserved name')
+
+    allocate(self % matIdxs(self % nMat))
+    if (dict % isPresent('all')) then
+      self % matIdxs = ALL_MATS
+    else
+      do i = 1, self % nMat
+        self % matIdxs(i) = mm_matIdx(mats(i))
+      end do
+    end if
+
+    ! Size field value array
+    self % outLocalID = product(self % sizeN) + 1
+    self % N = product(self % sizeN * self % nMat) + 1
+    allocate(self % val(self % N))
+
+    ! Read field values for each material
+    idx0 = 0
+    do i = 1, size(mats)
+
+      call dict % get(temp, mats(i))
+
+      ! Flip array up-down for more natural input
+      ! Reshape into rank 2 array
+      tempMap = reshape(temp, [self % sizeN(1), self % sizeN(2) * self % sizeN(3)])
+      N = size(tempMap, 2)
+      do j = 1, N/2
+        call swap(tempMap(:,j), tempMap(:,N - j + 1))
```

Comment:
It's not quite clear why you want to swap the array. Also, couldn't you make val 3D for better interpretability?